osh/builtin_comp.py

```diff
@@ -129,22 +143,22 @@ def Build(self, argv, attrs, base_opts):
         # directory, and external commands in $PATH.
 
         actions.append(_FixedWordsAction(consts.BUILTIN_NAMES))
-        actions.append(_DynamicDictAction(self.parse_ctx.aliases))
-        actions.append(_DynamicDictAction(cmd_ev.procs))
+        actions.append(_DynamicStrDictAction(self.parse_ctx.aliases))
+        actions.append(_DynamicProcDictAction(cmd_ev.procs))
         actions.append(_FixedWordsAction(consts.OSH_KEYWORD_NAMES))
-        actions.append(completion.FileSystemAction(exec_only=True))
+        actions.append(completion.FileSystemAction(False, True, False))
 
         # Look on the file system.
         a = completion.ExternalCommandAction(cmd_ev.mem)
 
       elif name == 'directory':
-        a = completion.FileSystemAction(dirs_only=True)
+        a = completion.FileSystemAction(True, False, False)
 
       elif name == 'file':
-        a = completion.FileSystemAction()
+        a = completion.FileSystemAction(False, False, False)
 
       elif name == 'function':
-        a = _DynamicDictAction(cmd_ev.procs)
+        a = _FixedWordsAction(cmd_ev.procs.keys())
```

Comment:
Oh I think you can remove `.keys()` here --  this is going to have the bug where when `procs`  is updated with new procs, they won't be completed

Reply:
Ah nice catch! This was leftover from before we talked about the difference between this and FixedWordsAction. Will fix